File: python/oecd/oecd_function.py
from __future__ import annotations

import io
import pandas as pd
import requests


def oecd_api_function(
    agency_identifier: str,
    dataset_identifier: str,
    data_selection: str,
    base_url: str = "https://sdmx.oecd.org/public/rest/data",
    dataset_version: str = "",
    startPeriod: str | None = None,
    endPeriod: str | None = None,
    dimensionAtObservation: str | None = None,
) -> pd.DataFrame:
    """
    Descargar datos de la OCDE (SDMX CSV) y devolver DataFrame.
    """
    data_identifier = f"{agency_identifier},{dataset_identifier},{dataset_version}"
    params: dict[str, str] = {}
    if startPeriod:
        params["startPeriod"] = startPeriod
    if endPeriod:
        params["endPeriod"] = endPeriod
    if dimensionAtObservation:
        params["dimensionAtObservation"] = dimensionAtObservation

    url = f"{base_url}/{data_identifier}/{data_selection}"
    resp = requests.get(url, params=params, headers={"Accept": "text/csv"}, timeout=180)
    resp.raise_for_status()

    text_stream = io.StringIO(resp.content.decode("utf-8"))
    df = pd.read_csv(text_stream)
    return df

File: R/oecd/oecd_function.R
# Función mínima para descargar datos de OECD SDMX (CSV)

# Inputs obligatorios:
# - agency_identifier: identificador de la agencia, p. ej. "OECD.ECO.MAD"
# - dataset_identifier: identificador del dataset, p. ej. "DSD_EO@DF_EO"
# - data_selection: clave SDMX (dimensiones) tras la '/'
#
# Inputs opcionales:
# - base_url: host de la API SDMX
# - dataset_version: versión del dataset
# - startPeriod, endPeriod, dimensionAtObservation: parámetros comunes de consulta
#
# Output:
# - df: data.frame con los datos descargados

oecd_api_function <- function(
  agency_identifier,
  dataset_identifier,
  data_selection,
  base_url = "https://sdmx.oecd.org/public/rest/data",
  dataset_version = "",
  startPeriod = NULL,
  endPeriod = NULL,
  dimensionAtObservation = NULL
) {
  data_identifier <- paste0(agency_identifier, ",", dataset_identifier, ",", dataset_version)

  params <- list()
  if (!is.null(startPeriod)) params$startPeriod <- startPeriod
  if (!is.null(endPeriod)) params$endPeriod <- endPeriod
  if (!is.null(dimensionAtObservation)) params$dimensionAtObservation <- dimensionAtObservation

  url <- paste0(base_url, "/", data_identifier, "/", data_selection)
  res <- httr::GET(url, query = params, httr::accept("text/csv"))
  if (httr::http_error(res)) stop(sprintf("OECD request failed [%s]", httr::status_code(res)))

  txt <- httr::content(res, as = "text", encoding = "UTF-8")
  df <- utils::read.csv(text = txt, stringsAsFactors = FALSE)
  return(df)
}


